When loading incompatible savestate, display exact Dolphin version

JosJuice opened the pull request:
Sometimes people don't remember what version of Dolphin was used to create a savestate. This PR makes Dolphin save its version as a string to all savestates it creates, and makes Dolphin read the version string from all savestates that were created after this feature was added. In case the user loads an incompatible savestate, a panic alert is displayed that tells the user which version of Dolphin was used to create the savestate, so that the user will be able to find that version and load the savestate.

The old "Can't load state from other revisions !" OSD message is still displayed, in addition to the new panic alert. Should the OSD message be kept, or is it better to only have the new one?

lioncash commented on Source/Core/Core/State.cpp:
```diff
@@ -90,10 +91,25 @@ static void DoState(PointerWrap &p)
+	{
+		p.Do(version_string);
+	}
+	else
```
The braces here aren't necessary.

skidau commented:
Please change the PanicAlert to a OSD.  PanicAlerts are to be used if Dolphin is about to crash to desktop.

JosJuice commented:
The problem with OSD messages is that you can't stop and read for as long as you want to. I guess I could try to cram the version number in while keeping the message short, but users probably want to have the version number on screen instead of having to memorize a four-digit number, and I also had a request about making the panic alert explain that a proper save should be used instead of a savestate. So while this isn't exactly the perfect way to use a panic alert, I don't think there's a better option.

skidau commented:
The thing with PanicAlerts is that the user has to click the mouse button or hit the enter key on the keyboard to continue the game.  This is inconvenient if they are (not near the kb/m) playing from a gamepad or wiimote.

JosJuice commented:
Yes, but it's much less inconvenient than having to exit Dolphin, potentially download a separate version and then run that. I don't think users try to load old savestates in the middle of gameplay often either... unless they do it by accident when trying to load another savestate? Is that a concern?

JosJuice commented:
I have now also added support for displaying approximate Dolphin versions when loading older savestates.

Oddly enough, I noticed that the panic alerts don't show up if I disable OSD messages, even if Use Panic Handlers are turned on. What's up with that?

JosJuice commented on Source/Core/Core/State.cpp:
```diff
@@ -64,7 +65,39 @@ static Common::Event g_compressAndDumpStateSyncEvent;
+	{ 38, { "4.0-4963", "4.0-5267" } },
+	{ 39, { "4.0-5279", "4.0-5525" } },
+	{ 40, { "4.0-5531", "4.0-5809" } },
+	{ 41, { "4.0-5811", "4.0-????" } }
```
This line needs to be updated right before this PR is merged.

JosJuice commented on Source/Core/Core/State.cpp:
```diff
@@ -64,7 +65,40 @@ static Common::Event g_compressAndDumpStateSyncEvent;
+	{ 39, { "4.0-5279", "4.0-5525" } },
+	{ 40, { "4.0-5531", "4.0-5809" } },
+	{ 41, { "4.0-5811", "4.0-5923" } },
+	{ 42, { "4.0-5925", "4.0-????" } }
```
This line needs to be updated right before this PR is merged.

RachelBryk commented:
I am not sure it is safe to assume people will always go download the old version to load their old save state rather than just give up and not use it. -1 on panicalert.

JosJuice commented:
Okay. I added it beacuse of those people who use savestates as their only method of saving, but I suppose there are a lot more users that only use savestates lightly and just do what you said. I've removed the second part of the message now, so that it only says that the savestate is incompatible and which version it was created with.

Was the "-1 on panicalert" only referring to the text in the panic alert, or the fact that there is a panic alert at all?

RachelBryk commented:
The fact that there is one at all. It should be an OSD message.

JosJuice commented:
Then OSD it is. There is now a second OSD message that displays the version, in addition to the existing one that says "Can't load state from other versions!"
